### PR TITLE
Fix/osc

### PR DESCRIPTION
**Pull request by cremebrule**

### CODE REVIEW

**Review by wensi-ai — APPROVED** [review]
lgtm!

File: omnigibson/robots/manipulation_robot.py [hunk 1]
```diff
@@ -417,7 +417,7 @@ def _add_arm_control_dict(self, fcns, arm):
         eef_link_idx = self._articulation_view.get_body_index(self.eef_links[arm].body_name)
         fcns[f"eef_{arm}_jacobian_relative"] = lambda: ControllableObjectViewAPI.get_relative_jacobian(
             self.articulation_root_path
-        )[eef_link_idx, :, start_idx : start_idx + self.n_joints]
+        )[-(self.n_links - eef_link_idx), :, start_idx : start_idx + self.n_joints]
```
Comment (ChengshuLi) [inline omnigibson/robots/manipulation_robot.py:420]: sorry can you add a comment explaining why the following line?
```
-(self.n_links - eef_link_idx)
```

File: omnigibson/utils/usd_utils.py [hunk 2]
```diff
@@ -947,7 +961,7 @@ def get_relative_linear_velocity(self, prim_path):
     def get_relative_angular_velocity(self, prim_path):
         orn = self.get_position_orientation(prim_path)[1]
         angvel = self.get_angular_velocity(prim_path)
-        return T.mat2euler(T.quat2mat(orn).T @ T.euler2mat(angvel))
+        return T.quat2mat(orn).T @ angvel
```
Comment (ChengshuLi) [inline omnigibson/utils/usd_utils.py:964]: this `T.xxx.T` syntax is very confusing lol

File: tests/test_controllers.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        # y should be positive
+        return (curr_position[1] - init_position[1]).item() > side_tol and th.norm(
+            curr_position[[0, 2]] - init_position[[0, 2]]
+        ).item() < tol
+
+    def check_up_error(curr_position, init_position, tol=1e-2, up_tol=1e-2):
+        # z should be positive
+        return (curr_position[2] - init_position[2]).item() > up_tol and th.norm(
+            curr_position[[0, 1]] - init_position[[0, 1]]
+        ).item() < tol
```
Comment (ChengshuLi) [inline tests/test_controllers.py:60]: Is it possible to do a quantitative check, rather than a qualitative test?
e.g. command the eef to move 5cm and confirm it is the case.

File: tests/test_controllers.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+                        f"curr_pos: {curr_pos}, init_pos: {init_pos}"
+                    )
+                    ori_err_normalized = th.norm(
+                        T.quat2axisangle(T.mat2quat(T.quat2mat(init_quat).T @ T.quat2mat(curr_quat)))
+                    ).item() / (np.pi * 2)
+                    ori_err = np.abs(np.pi * 2 * (np.round(ori_err_normalized) - ori_err_normalized))
+                    assert ori_err < 0.1, (
+                        f"Got mismatch for controller [{controller}], robot [{robot.model_name}], action [{action_name}]\n"
+                        f"curr_quat: {curr_quat}, init_quat: {init_quat}, err: {ori_err}"
+                    )
```
Comment (ChengshuLi) [inline tests/test_controllers.py:165]: can we do a quantitative test for orientation as well? e.g. command the eef to rotate 45 degrees and confirm it's the case.

File: tests/test_controllers.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
+import numpy as np
+import pytest
+import torch as th
+
+import omnigibson as og
+import omnigibson.utils.transform_utils as T
+from omnigibson.robots import LocomotionRobot
+
+
+def test_arm_control():
```
Comment (ChengshuLi) [inline tests/test_controllers.py:10]: You should also consider taking a look at Hang's test in `test_robot_states_flatcache.py`

**Review by ChengshuLi — APPROVED** [review]
LGTM with some minor comments! Feel free to merge once those are addressed.

File: omnigibson/robots/manipulation_robot.py (hunk 1 above)
Comment (cremebrule) [reply omnigibson/robots/manipulation_robot.py:420]: done!

File: omnigibson/utils/usd_utils.py (hunk 2 above)
Comment (cremebrule) [reply omnigibson/utils/usd_utils.py:964]: haha it's aesthetic because it's symmetrical ;D

Comment (cremebrule) [reply omnigibson/utils/usd_utils.py:964]: added a comment for clarity

File: tests/test_controllers.py (hunk 4 above)
Comment (cremebrule) [reply tests/test_controllers.py]: done!

File: tests/test_controllers.py (hunk 3 above)
Comment (cremebrule) [reply tests/test_controllers.py]: done!

File: omnigibson/controllers/ik_controller.py [hunk 6]
```diff
@@ -135,19 +137,28 @@ def limiter(target_pos: Array[float], target_quat: Array[float], control_dict: D
             else MovingAverageFilter(obs_dim=control_dim, filter_width=smoothing_filter_size)
         )
         assert mode in IK_MODES, f"Invalid ik mode specified! Valid options are: {IK_MODES}, got: {mode}"
+
+        # If mode is absolute pose, make sure command input limits / output limits are None
```
Comment (ChengshuLi) [inline omnigibson/controllers/ik_controller.py:141]: could you add a comment explaining why they should be None?

File: omnigibson/robots/manipulation_robot.py [hunk 7]
```diff
@@ -413,11 +413,14 @@ def _add_arm_control_dict(self, fcns, arm):
         # -n_joints because there may be an additional 6 entries at the beginning of the array, if this robot does
         # not have a fixed base (i.e.: the 6DOF --> "floating" joint)
         # see self.get_relative_jacobian() for more info
+        # We also count backwards for the link frame because if the robot is fixed base, the jacobian returned has one
+        # less index than the number of links. This is presumably because the 1st link of a fixed base robot will
+        # always have a zero jacobian since it can't move. Counting backwards resolves this issue.
```
Comment (ChengshuLi) [inline omnigibson/robots/manipulation_robot.py:418]: wow is this the concrete bug you found?

File: omnigibson/utils/transform_utils.py [hunk 8]
```diff
@@ -697,7 +697,7 @@ def axisangle2quat(vec, eps=1e-6):
     angle = th.norm(vec, dim=-1, keepdim=True)
 
     # Create return array
-    quat = th.zeros(th.prod(th.tensor(input_shape)), 4, device=vec.device)
+    quat = th.zeros(th.prod(th.tensor(input_shape, dtype=th.int)), 4, device=vec.device)
```
Comment (ChengshuLi) [inline omnigibson/utils/transform_utils.py:700]: why are these `dtype` needed?

**Review by ChengshuLi — APPROVED** [review]

File: omnigibson/robots/manipulation_robot.py (hunk 7 above)
Comment (cremebrule) [reply omnigibson/robots/manipulation_robot.py:418]: well this is another bug found, but i don't think this caused the major issue

File: omnigibson/controllers/ik_controller.py (hunk 6 above)
Comment (cremebrule) [reply omnigibson/controllers/ik_controller.py:141]: It doesn't make sense for the action to be rescaled, if you're assuming the input is an absolute value -- it sort of defeats the purpose if you ask for a specific absolute pose and it gets rescaled arbitrarily. This was a silent bug that was occurring where by default actions are normalized, and therefore actions are rescaled. And so the unit test with absolute_pose was failing for this reason

File: omnigibson/utils/transform_utils.py (hunk 8 above)
Comment (cremebrule) [reply omnigibson/utils/transform_utils.py:700]: oh it's just to pacify some torch warnings that were being printed about the dtype being inferred. Better to be explicit than implicit, I assume?
